Port V2 API calls

SimonRichardson opened the pull request:
Adding the ability to create, get, list and delete a series of ports
for the neutron API.

The following code duplicates most of the existing work found with in
the neutron networking code, but adapts it for Ports. The API
documentation for the Ports API can be found[1].

This is required to provide spaces and multiple nic support to Juju. See
work in progress PR[2]

1. https://docs.openstack.org/api-ref/network/v2/index.html?expanded=create-port-detail#ports
2. https://github.com/juju/juju/pull/11188

SimonRichardson committed:
Port V2 API calls

Adding the ability to create, get, list and delete a series of ports
for the neutron API.

The following code duplicates most of the existing work found with in
the neutron networking code, but adapts it for Ports. The API
documentation for the Ports API can be found[1].

This is required to provide spaces and multiple nic support to Juju. See
work in progress PR[2]

1. https://docs.openstack.org/api-ref/network/v2/index.html?expanded=create-port-detail#ports
2. https://github.com/juju/juju/pull/11188

manadart reviewed (approved):
Nice.

SimonRichardson committed:
Adding neutronservice tests

The following starts to add stubbing test code, along with tests for
that test code.

Turtles all the way down, for an infinite regress.

SimonRichardson committed:
Add ports http service testing

The following adds tests to ensure that the http server works as
expected. The feels a lot like testing the tests at which point you have
to ask if this is overkill?

I feel in the future it would be better served to write mocks for the
API and not have so much http testing tests. One has to realise that
you're actually testing the tests and not the integration with the
openstack API.

SimonRichardson committed:
Test delete directly

The following tests delete directly so that we can ensure that the API
is correctly wired up.

Also moved to url param instead of a query param for the get method.

hmlanigan commented on neutron/neutron.go:
```diff
@@ -231,6 +233,85 @@ func (c *Client) DeleteFloatingIPV2(ipId string) error {
 	return err
 }
 
+// PortV2 describes a defined network for creating a port.
```
PortV2 describes a Port, used for more than creating a port.

hmlanigan commented on neutron/neutron.go:
```diff
@@ -231,6 +233,85 @@ func (c *Client) DeleteFloatingIPV2(ipId string) error {
+	NetworkId      string           `json:"network_id"`
+	SecurityGroups []string         `json:"security_groups,omitempty"`
+	Status         string           `json:"status,omitempty"`
+	TenantId       string           `json:"tenant_id,omitempty"`
```
"port_security_enabled" might be useful too.  I know we need it for the networks, not sure if it can be port specific or not.

"device_id" might be useful too, for mapping to server instances.

hmlanigan commented on neutron/neutron.go:
```diff
@@ -231,6 +233,85 @@ func (c *Client) DeleteFloatingIPV2(ipId string) error {
+}
+
+// ListPortsV2 lists NetworkIds, names, and other details for all ports.
+func (c *Client) ListPortsV2() ([]PortV2, error) {
```
Suggest you add Filter to this, like used for ListNetworks and ListFloatingIPs.  Might be nice later.

hmlanigan commented on neutron/neutron.go:
```diff
@@ -231,6 +233,85 @@ func (c *Client) DeleteFloatingIPV2(ipId string) error {
+	TenantId       string           `json:"tenant_id,omitempty"`
+}
+
+// PortFixedIPsV2 represents a FIxedIp with ip addresses and an associated
```
s/FIx/Fix/

hmlanigan reviewed (approved):
A few suggestions for things to add, LTGM.

SimonRichardson committed:
Add List filtering

Adding list filtering is an expectation when using the goose library and
it has it in other places. The server code was setup for this, but the
library code wasn't able to pass that in. Adding the filtering was
really simple by just passing in the generic filter option.

Additionally fixing typo and ensuring we're exposing everything we need
when requesting the port.
